### PR TITLE
fix(http-client-csharp): remove setters for model properties collection types

### PR DESCRIPTION
**Pull request by jorgerangel-msft**
This PR fixes the case where setters were being generated for model properties that were collection types. Model properties that are collection types will no longer generated a setter unless they are nullable.

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs [hunk 1]
```diff
@@ -48,21 +49,41 @@ protected override PropertyDeclaration[] BuildProperties()
...
+                return false;
+            }
+
+            if (type.IsCollection)
+            {
+                return type.IsNullable;
+            }
+
+            return true;
+        }
```
Comment (ShivangiReja) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs:85]: In autorest.csharp, we have all these conditions in [this file](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs#L195-L232) that determine whether we need a setter for a property or not.

Particularly for collections, we have [this](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs#L224C17-L224C60) condition: `type.IsCollection && !type.IsReadOnlyMemory`. I believe we need to include the condition `type` is not `ReadOnlyMemory` here, especially in the case for vector properties.

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs]: thanks. I've pulled in some of these additional conditions. Do let me know if you still think the conditions were adding aren't enough to cover all the properties we need to generate.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs [hunk 2]
```diff
@@ -48,21 +49,56 @@ protected override PropertyDeclaration[] BuildProperties()
         private PropertyDeclaration BuildPropertyDeclaration(InputModelProperty property)
         {
             var propertyType = CodeModelPlugin.Instance.TypeFactory.CreateCSharpType(property.Type);
+            var propHasSetter = PropertyHasSetter(propertyType, property);
             MethodSignatureModifiers setterModifier = !property.IsReadOnly ? MethodSignatureModifiers.Public : MethodSignatureModifiers.None;
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs:53]: should we change this condition here as well?
```suggestion
            MethodSignatureModifiers setterModifier = propHasSetter ? MethodSignatureModifiers.Public : MethodSignatureModifiers.None;
```

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel.TestProjects/Unbranded-TypeSpec/src/Generated/Models/Thing.cs [hunk 3]
```diff
@@ -27,7 +27,7 @@ public partial class Thing
         public string RequiredLiteralFloat { get; set; } = "1.23";
```
Comment (ShivangiReja) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel.TestProjects/Unbranded-TypeSpec/src/Generated/Models/Thing.cs:27]: The IsLiteral property of CSharpType is incorrectly returning false for int and float. We have a issue to track this - https://github.com/Azure/autorest.csharp/issues/4630.

Comment (ArcturusZhang) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel.TestProjects/Unbranded-TypeSpec/src/Generated/Models/Thing.cs:27]: I am trying to fix it in this PR: https://github.com/microsoft/typespec/pull/3224 (despite its name does not mention this, it is still WIP)

**Review by ArcturusZhang — APPROVED** [review]

**Review by ShivangiReja — APPROVED** [review]

**Pull request merged**